File: Humming/Core/Pitch.swift
import Foundation

public enum Pitch {
    public static let minHz: Double = 70
    public static let maxHz: Double = 900

    public static func hzToMidi(_ hz: Double) -> Double {
        69 + 12 * log2(hz / 440)
    }

    public static func midiToHz(_ midi: Double) -> Double {
        440 * pow(2, (midi - 69) / 12)
    }

    private static let noteNames = ["C", "C#", "D", "D#", "E", "F", "F#", "G", "G#", "A", "A#", "B"]

    public static func midiToNoteName(_ midi: Double) -> String {
        let rounded = Int(midi.rounded())
        let name = noteNames[((rounded % 12) + 12) % 12]
        let octave = Int(floor(Double(rounded) / 12.0)) - 1
        return "\(name)\(octave)"
    }

    public static func sineWave(hz: Double, sampleRate: Double, seconds: Double, amplitude: Float = 0.6) -> [Float] {
        let n = Int(sampleRate * seconds)
        return (0..<n).map { i in
            Float(sin(2 * Double.pi * hz * Double(i) / sampleRate)) * amplitude
        }
    }

    /// YIN pitch detection. Returns Hz, or nil if unvoiced / out of humming range.
    public static func detect(_ buffer: [Float], sampleRate: Double, threshold: Float = 0.15) -> Double? {
        let half = buffer.count / 2
        guard half >= 32 else { return nil }

        var yin = [Float](repeating: 0, count: half)
        for tau in 1..<half {
            var sum: Float = 0
            for i in 0..<half {
                let delta = buffer[i] - buffer[i + tau]
                sum += delta * delta
            }
            yin[tau] = sum
        }

        yin[0] = 1
        var running: Float = 0
        for tau in 1..<half {
            running += yin[tau]
            yin[tau] = running == 0 ? 1 : (yin[tau] * Float(tau)) / running
        }

        let minTau = max(2, Int(sampleRate / maxHz))
        let maxTau = min(half - 1, Int(sampleRate / minHz))

        var tau = minTau
        while tau <= maxTau {
            if yin[tau] < threshold {
                while tau + 1 <= maxTau, yin[tau + 1] < yin[tau] {
                    tau += 1
                }
                break
            }
            tau += 1
        }

        guard tau <= maxTau, yin[tau] < threshold else { return nil }

        let prev = tau > 0 ? yin[tau - 1] : 1
        let next = tau + 1 < yin.count ? yin[tau + 1] : 1
        let denom = 2 * (2 * yin[tau] - prev - next)
        let shift = denom == 0 ? 0 : Double((prev - next) / denom)
        let freq = sampleRate / (Double(tau) + shift)
        guard freq.isFinite, freq >= minHz, freq <= maxHz else { return nil }
        return freq
    }
}

File: src/audio/pitch.ts
const HUM_MIN_HZ = 70
const HUM_MAX_HZ = 900

export function hzToMidi(hz: number): number {
  return 69 + 12 * Math.log2(hz / 440)
}

export function midiToHz(midi: number): number {
  return 440 * 2 ** ((midi - 69) / 12)
}

const NOTE_NAMES = ['C', 'C#', 'D', 'D#', 'E', 'F', 'F#', 'G', 'G#', 'A', 'A#', 'B'] as const

export function midiToNoteName(midi: number): string {
  const rounded = Math.round(midi)
  const name = NOTE_NAMES[((rounded % 12) + 12) % 12]
  const octave = Math.floor(rounded / 12) - 1
  return `${name}${octave}`
}

function parabolicPeak(prev: number, current: number, next: number): number {
  const denom = 2 * (2 * current - prev - next)
  if (denom === 0) return 0
  return (prev - next) / denom
}

/**
 * YIN pitch detection. Returns Hz or null if unvoiced / out of humming range.
 */
export function detectPitch(
  buffer: Float32Array,
  sampleRate: number,
  threshold = 0.15,
): number | null {
  const half = Math.floor(buffer.length / 2)
  if (half < 32) return null

  const yin = new Float32Array(half)
  for (let tau = 1; tau < half; tau++) {
    let sum = 0
    for (let i = 0; i < half; i++) {
      const delta = buffer[i]! - buffer[i + tau]!
      sum += delta * delta
    }
    yin[tau] = sum
  }

  yin[0] = 1
  let running = 0
  for (let tau = 1; tau < half; tau++) {
    running += yin[tau]!
    yin[tau] = running === 0 ? 1 : (yin[tau]! * tau) / running
  }

  const minTau = Math.max(2, Math.floor(sampleRate / HUM_MAX_HZ))
  const maxTau = Math.min(half - 1, Math.floor(sampleRate / HUM_MIN_HZ))

  let tau = minTau
  for (; tau <= maxTau; tau++) {
    if (yin[tau]! < threshold) {
      while (tau + 1 <= maxTau && yin[tau + 1]! < yin[tau]!) tau++
      break
    }
  }

  if (tau > maxTau || yin[tau]! >= threshold) return null

  const shift = parabolicPeak(yin[tau - 1] ?? 1, yin[tau]!, yin[tau + 1] ?? 1)
  const freq = sampleRate / (tau + shift)
  if (!Number.isFinite(freq) || freq < HUM_MIN_HZ || freq > HUM_MAX_HZ) return null
  return freq
}

export function sineWave(
  hz: number,
  sampleRate: number,
  seconds: number,
  amplitude = 0.6,
): Float32Array {
  const n = Math.floor(sampleRate * seconds)
  const out = new Float32Array(n)
  for (let i = 0; i < n; i++) {
    out[i] = Math.sin((2 * Math.PI * hz * i) / sampleRate) * amplitude
  }
  return out
}
